kep/NNN-preemption-toleration/README.md

```diff
@@ -0,0 +1,125 @@
+
+Especially, this plugin enables cluster administrators to define preemption policy from the _victim_ side which is not covered public `PriorityClass` API. This plugin provides _preemption toleration_ policy in `PriorityClass`.  The policy defines the criteria by which the priority class will be exempt from preemption.
+
+## Motivation
+
+Kubernetes scheduler provides [Pod Priority and Preemption](https://kubernetes.io/docs/concepts/scheduling-eviction/pod-priority-preemption/) feature.  Users, typically cluster administrators, can define several `PriorityClass` to classify priority in the cluster.  If the cluster is less elastic (e.g. On-Premise clusters), designing priority classes and preemption rules are very important for computational resource utilization.
+
+`PriorityClass` can have `PreemptionPolicy` configuration for customizing preemptor behavior for the priority class. The allowed values are `PreemptLowerPriority`(default) and `Never`.  If `Never` is set, the priority class becomes [_Non-preempting priority class_](https://kubernetes.io/docs/concepts/scheduling-eviction/pod-priority-preemption/#non-preempting-priority-class) that can not preempt any pods but may be preempted by higher priority pods.  The `PreemptionPolicy` API is very simple and understandable.  However, this policy only focuses on preemptor side behavior.
+
+This plugin provides more flexible preemption behavior configuration by adding the preemptee (a.k.a victim) side policy in `PriorityClass`. In particular, cluster administrators can define _preemption toleration_ policy that defines the criteria which the priority class will be exempt from preemption.
+
+### Goals
+
+- provide flexible preemption toleration policy spec
+- and its implementation
+ 
+### Non-Goals
+
+- change Kubernetes public `PriorityClass` API
+
+## Use Cases
+
+### Guarantee to running at least N minutes even in lower priority
+
+This is a typical use-case in machine learning job. To make a machine learning job be resumable after preemption, the job need to preserve its state to outside of the pod (e.g. persistent volume, object storage) at preemption and restore the state at next restart. However, most machine learning jobs are iterative based on epochs. So, if preemption always happens before the first epoch is finished,  the machine learning jobs can never make any progress and consumed computational resources came to nothing in this case.
+
+So, cluster administrators would like to exempt pods that do not finish the first epoch from preemption even if the job has lower priority.
+
+However, it is hard to monitor when the first epoch is finished outside of the job. Thus, a minimum running time guarantee is useful practically.  For example, cluster administrator would introduce several priority classes with the same priority value for introducing several minimum running time guarantee categories like below:
```

Comment:
Does the "running time" include pods' initialization phase (pulling images, running init containers, etc) or, only include duration after pods reach the `Running` phase?
We should determine an exact definition and avoid the term "minimumRunningDuration" because it could be confusing.

Reply:
I think we could use time from being scheduled.